### PR TITLE
Make runtime and trap errors well defined

### LINKED ISSUE #1328: RuntimeError is not always downcastable to the host error
### Describe the bug

In Rust, sometimes `wasmer_runtime::error::RuntimeError` returned by Wasmer is not downcast-able to the error type that our host function returns to Wasmer.

* Here is we try to downcast it:
https://github.com/nearprotocol/nearcore/blob/a08944db6672185618e2454fce5e7be0e327b47a/runtime/near-vm-runner/src/errors.rs#L59
* And here is the place where all host functions return only `VMLogicError `https://github.com/nearprotocol/nearcore/blob/05a82d8dd74531b21567623ab0e7e75946afea19/runtime/near-vm-runner/src/imports.rs#L7

Versions:
```
wasmer-runtime = { version = "0.11.0", features = ["default-backend-singlepass"], default-features = false }
wasmer-runtime-core = { version = "0.11.0", features = ["backend-singlepass"]}
```

### Expected behavior
`RuntimeError` to be always downcastable to the error returned by the host functions.

### Actual behavior
Sometimes we are getting error that cannot be downcasted.

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
Resolves #1328 

This PR goes through and gives explicit types for all the errors instead of using `Box<dyn Any + Send>` where possible.  This gives users better insight into what the specific errors are and should help with debugging in the case of mistakes in our code.

The remaining uses of `Box<dyn Any>` are due to the structure of our dependency graph -- this is probably solvable but it seems fine as is as all error types are now explicit and the remaining `Box<dyn Any>`s are either fully user controlled or not for end-user consumption.

# Review

- [x] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try-

**Comment by MarkMcCaskey** [comment]
bors try

**Comment by MarkMcCaskey** [comment]
bors try

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 ## **[Unreleased]**
 
+- [#1401](https://github.com/wasmerio/wasmer/pull/1401) Make breaking change to `RuntimeError`: `RuntimeError` is now more explicit about its possible error values allowing for better insight into why a call may have failed.
```
Comment (nlewycky) [inline CHANGELOG.md:5]: Suggest removing "may have". The usage of `may have` here could read as if the call may or may not have failed, I think you meant it as in the more explicit error may have more details, but that's already couched in the first half of the sentence.

File: examples/callback-guest/callback-guest.rs [hunk 2]
```diff
@@ -4,7 +4,6 @@ extern "C" {
     fn host_callback() -> u32;
 }
 
-#[no_mangle]
```
Comment (nlewycky) [inline examples/callback-guest/callback-guest.rs:7]: What's the motivation for this change?

File: lib/clif-backend/src/signal/mod.rs [hunk 3]
```diff
@@ -80,7 +79,10 @@ impl RunnableModule for Caller {
 
             match res {
                 Err(err) => {
-                    *error_out = Some(err.0);
+                    // probably makes the most sense to actually do a translation here to a
+                    // a generic type defined in runtime-core
+                    // TODO: figure out _this_ error return story
```
Comment (nlewycky) [inline lib/clif-backend/src/signal/mod.rs:84]: Did you mean to do this before sending out the PR, or to commit the TODO?

File: lib/clif-backend/src/signal/windows.rs [hunk 4]
```diff
@@ -32,14 +33,42 @@ thread_local! {
...
+        | EXCEPTION_FLT_OVERFLOW
+        | EXCEPTION_FLT_STACK_CHECK
+        | EXCEPTION_FLT_UNDERFLOW => "floating-point exception",
+        EXCEPTION_ILLEGAL_INSTRUCTION => "illegal instruction",
+        EXCEPTION_ACCESS_VIOLATION => "segmentation violation",
+        EXCEPTION_DATATYPE_MISALIGNMENT => "datatype misalignment",
+        EXCEPTION_BREAKPOINT => "breakpoint",
+        EXCEPTION_SINGLE_STEP => "single step",
+        EXCEPTION_ARRAY_BOUNDS_EXCEEDED => "array bounds exceeded",
+        EXCEPTION_INT_DIVIDE_BY_ZERO => "int div by zero",
```
Comment (nlewycky) [inline lib/clif-backend/src/signal/windows.rs:51]: Optional:
```suggestion
        EXCEPTION_INT_DIVIDE_BY_ZERO => "integer division by zero",
        EXCEPTION_INT_OVERFLOW => "integer overflow",
        EXCEPTION_PRIV_INSTRUCTION => "privileged instruction",
```

File: lib/llvm-backend/cpp/object_loader.hh [hunk 5]
```diff
@@ -49,9 +49,7 @@ typedef struct {
   visit_fde_t visit_fde;
 } callbacks_t;
 
-typedef struct {
-  size_t data, vtable;
-} box_any_t;
+using runtime_error_t = void*;
```
Comment (nlewycky) [inline lib/llvm-backend/cpp/object_loader.hh:52]:
```suggestion
typedef struct runtime_error_t_privdecl *runtime_error_t;
```
will get you a typesafe pointer-to-opaque, where the opaque is a forward declaration of a struct that is never defined.

However, maybe this type shouldn't be a pointer, and all the references to `runtime_error_t` should really be `runtime_error_t*`? You could write `struct runtime_error_t;` here and never provide it a body. Is that closer to what's really going on?

File: lib/runtime-core/src/error.rs [hunk 6]
```diff
@@ -173,43 +173,133 @@ impl std::fmt::Display for LinkError {
...
-/// The main way to do this is `Instance.call`.
+/// An error that happened while invoking a Wasm function.
+#[derive(Debug)]
+pub enum InvokeError {
+    /// Indicates an exceptional circumstance such as a bug in Wasmer (please file an issue!)
+    /// or a hardware failure.
+    FailedWithNoError,
+    /// Indicates that a trap occurred that is not known to Wasmer.
+    UnknownTrap {
+        /// The address that the trap occured at.
```
Comment (nlewycky) [inline lib/runtime-core/src/error.rs:184]:
```suggestion
        /// The address that the trap occurred at.
```

File: lib/runtime-core/src/error.rs [hunk 7]
```diff
@@ -173,43 +173,133 @@ impl std::fmt::Display for LinkError {
...
+/// `RuntimeError(Box::new(<string>))`.
 ///
-/// Comparing two `RuntimeError`s always evaluates to false.
-pub struct RuntimeError(pub Box<dyn Any + Send>);
+/// This type provides greater visibility into the kinds of things that may fail
+/// and improves the ability of users to handle them, though these errors may be
+/// extremely rare and impossible to handle.
+#[derive(Debug)]
+pub enum RuntimeError {
+    /// When an invoke returns an error
```
Comment (nlewycky) [inline lib/runtime-core/src/error.rs:254]:
```suggestion
    /// When an invoke returns an error.
```

File: examples/callback-guest/callback-guest.rs (hunk 2 above)
Comment (MarkMcCaskey) [reply examples/callback-guest/callback-guest.rs:7]: Oh that was not supposed to be part of this PR -- I was testing something for someone on our public slack

File: lib/clif-backend/src/signal/mod.rs (hunk 3 above)
Comment (MarkMcCaskey) [reply lib/clif-backend/src/signal/mod.rs:84]: Outdated comment -- already did it! Thanks!

File: lib/llvm-backend/cpp/object_loader.hh (hunk 5 above)
Comment (MarkMcCaskey) [reply lib/llvm-backend/cpp/object_loader.hh:52]: Hmm, I have mixed feelings about it, other than the explicit `copy` function exported from Rust, the C++ code doesn't need to know anything about this type... that copy function does change things though
